TS number tests with embedded study

guilpier-code opened the pull request.

guilpier-code committed:
Refactor TS numbers : refactor check intra-modal part

guilpier-code committed:
Refactor TS numbers : check intra-modal - fixing a bug + compilation problem

guilpier-code committed:
Refactor TS numbers : refactor ts number draws

guilpier-code committed:
[skip ci] Refactor TS numbers : a bit of cleaning

guilpier-code committed:
Refactor TS numbers : refactor check inter-modal correlations

guilpier-code committed:
[skip ci] Refactor TS numbers : a bit of cleaning

guilpier-code committed:
[skip ci] Refactor TS numbers : just enhance a message to users

guilpier-code committed:
Refactor TS numbers : factor intra-modal code by polymorphism

guilpier-code committed:
Refactor TS numbers : intra-modal -> bug fix + add polymorphism for no consistency message

guilpier-code committed:
[skip ci] Refactor TS numbers : intra-modal -> renaming

guilpier-code committed:
Refactor TS numbers : separate TS numbers draws steps inside functions

guilpier-code committed:
Refactor TS numbers : group inter-modal steps inside functions

guilpier-code committed:
Refactor TS numbers : renaming the classes retrieving the number of TS in an area for a given TS kind

guilpier-code committed:
Refactor TS numbers : not passing the whole study class when only a study component is needed.

guilpier-code committed:
Refactor TS numbers : inverting parameters in a function regarding intra-modal ts numbers storage

guilpier-code committed:
TS numbers - unit tests : remove useless headers inclusion

guilpier-code committed:
ts number unit tests : embed the whole study, 2 tests (one passes, one fails)

guilpier-code committed:
TS number unit tests (embed the whole study) : make the previous 2 tests pass

guilpier-code committed:
TS number unit tests (embed the whole study) : adding 2 tests on intra-modal with thermal clusters

guilpier-code committed:
TS number unit tests (embed the whole study) : adding 2 tests adding 2 tests on intra-modal with renewable clusters

flomnes commented on src/tests/src/solver/simulation/tests-ts-numbers.cpp:
```diff
@@ -0,0 +1,271 @@
+{}
+
+template<>
+void addClusterToAreaList(Area* area, std::shared_ptr<ThermalCluster> cluster)
+{
+	area->thermal.clusters.push_back(cluster.get());
+	area->thermal.list.add(cluster);
+	area->thermal.list.mapping[cluster->id()] = cluster;
+}
```
If the thermal cluster is "must-run", it should be added to mustRunList instead

flomnes replied:
~~l. 50 : `area->thermal.clusters` is of type `std::map<string, std::shared_ptr<ThermalCluster>>`. Therefore there is no need to call `get()` to access the raw pointer.~~

~~Doing so may lead to a double free (happened in the past).~~

I was wrong.

guilpier-code committed:
TS number unit tests (embed the whole study) : adding 1 test for inter-modal consistency

In this test :
- 1 area
- in this area : load, wind and thermal are in inter-modal

flomnes commented on src/tests/src/solver/simulation/tests-ts-numbers.cpp:
```diff
@@ -0,0 +1,305 @@
+	BOOST_CHECK(Generate(*study));
+}
+
+BOOST_AUTO_TEST_CASE(two_areas_3_thermal_clusters_with_different_number_of_ready_made_ts___check_intra_modal_ok)
```
Should be **ko**

flomnes reviewed (commented):
I mostly agree with this way to do things.

Add intermodal_ko tests ?

guilpier-code committed:
TS number unit tests (embed the whole study) : adding 3 tests for inter-modal consistency

flomnes commented on src/tests/src/solver/simulation/tests-ts-numbers.cpp:
```diff
@@ -0,0 +1,382 @@
+	area_2->load.series->series.resize(5, 1);
+
+	BOOST_CHECK(Generate(*study));
+}
```
Check that the TS numbers satisfy the intra-modal rule here.

guilpier-code replied:
done

flomnes commented on src/tests/src/solver/simulation/tests-ts-numbers.cpp:
```diff
@@ -0,0 +1,382 @@
+}
+
+BOOST_AUTO_TEST_CASE(two_areas_3_thermal_clusters_with_different_number_of_ready_made_ts___check_intra_modal_OK)
+{
```
KO

guilpier-code replied:
done

guilpier-code committed:
TS number unit tests (embed the whole study) : adding checks on stored ts numbers in intra/inter modal
- Inside an area, stored ts numbers of all inter-modal energy are equal
- Over all areas, stored ts numbers of any intra-modal energy are equal

guilpier-code committed:
TS number unit tests (embed the whole study) : test that drawn ts numbers are bounded (>=0 && < number of TS)

flomnes commented on src/tests/src/solver/simulation/CMakeLists.txt:
```diff
@@ -0,0 +1,53 @@
+						PUBLIC
+							libantares-core
+							libmodel_antares
+)
```
Why create a lib ? We could just add `libantares-core` and `libmodel_antares` after line 41 instead.

guilpier-code replied:
I wanted to separate sources from code and tests sources. 
Maybe not relevant.
Changes were done here.

flomnes commented on src/tests/src/solver/simulation/tests-ts-numbers.cpp:
```diff
@@ -0,0 +1,518 @@
+	BOOST_TEST((windTsNumber >= 0 && windTsNumber < windNumberOfTs));
+	BOOST_TEST((solarTsNumber>= 0 && solarTsNumber < solarNumberOfTs));
+	BOOST_TEST((hydroTsNumber >= 0 && hydroTsNumber < hydroNumberOfTs));
+	BOOST_TEST((thermalTsNumber>= 0 && thermalTsNumber < thermalNumberOfTs));
```
Remove useless parentheses.

guilpier-code replied:
done

guilpier-code committed:
[skip ci] Refactor TS numbers : replace (and fix) BUILD_LOG_ENTRY macro with a function

guilpier-code committed:
[skip ci] Refactor TS numbers : replace macro TS_INDEX with map ts_to_tsIndex

guilpier-code committed:
Refactor TS numbers : renaming + loop instead of several blocks of instructions

flomnes commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -63,19 +54,22 @@ const map<TimeSeries, string> ts_to_tsTitle = {
-            e += TEXT;               \
-        }                            \
-    } while (0)
+void addInterModalTimeSeriesToMessage(const bool* isTSintermodal, std::string & interModalTsMsg)
```
`std::array<bool, size>`

guilpier-code replied:
done

guilpier-code committed:
[skip ci] Refactor TS numbers : put TS numbers (one for each year) to zero when mode derated is ON

Indeed, derated mode implies that only one year is played, and its index is 0.

guilpier-code committed:
[skip ci] Refactor TS numbers : move array isTSintramodal from bool* to an std::array<bool, size>

flomnes commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -689,22 +666,19 @@ bool TimeSeriesNumbers::Generate(Data::Study& study)
+    const bool isTSgenerated[timeSeriesCount]
       = {0 != (Data::timeSeriesLoad & parameters.timeSeriesToRefresh),
          0 != (Data::timeSeriesHydro & parameters.timeSeriesToRefresh),
          0 != (Data::timeSeriesWind & parameters.timeSeriesToRefresh),
```
Use `std::array` for `nbTimeseriesByMode`, `intramodal_draws` and `isTSgenerated` as well ?

guilpier-code replied:
done

guilpier-code committed:
[skip ci] Refactor TS numbers : move C-style array into std::array, remove useless () and name space ref ("Data::")

guilpier-code committed:
TS number unit tests : clean CMakeLists.txt and tests

flomnes committed:
Format timeseries-numbers.cpp
